### PR TITLE
[framework] Add the ability to deprecate input and output ports

### PR DESCRIPTION
**Pull request by jwnimmer-tri**
Towards #12214.  This adds the _ability_ to deprecate ports in the first place.

No sugar for renaming input ports is implemented here, so I'll leave the issue open for now.

As a showcase of the feature, this deprecates MultibodyPlant's `continuous_state` port names, in lieu of just `state`.

This PR also adds an attorney for PortBase to de-clutter the public API.

---

When a port is marked as deprecated, the first user-induced attempt to use it within a process will result in a warning printed to the log.

Uses are detected as:

- Obtaining a reference to the InputPort or OutputPort handle (other than at declaration-time), with lookup either by name or index.

  Using the handle after lookup does not generate additional warnings.

  Note that connecting or exporting subsystem ports during Diagram construction counts as a detected use.

- Checking whether a port name exists.

- Obtaining the input port's cache ticket via the SystemBase.

- Evaluating an input port via the (dispreferred) direct System functions EvalAbstractInput or EvalInputValue.

- - -
This change is [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/18274)

### CODE REVIEW

**Comment by jwnimmer-tri** [comment]
To check for false positives, we can run:
`find -L bazel-testlogs -name test.log | xargs -n9999 grep -e 'OutputPort.*is deprecated'`.

Here's an example of a true positive from `multibody_plant_deprecated_test`:

```
[warning] OutputPort[2] (continuous_state) of System ::_ (MultibodyPlant<double>) is deprecated: Use the port named 'state' instead; the 'continuous_state' alias will be removed on or after 2023-03-01.
[warning] OutputPort[11] (DefaultModelInstance_continuous_state) of System ::_ (MultibodyPlant<double>) is deprecated: Use the port suffxed with '_state' instead; the '_continuous_state' alias will be removed on or after 2023-03-01.
[warning] OutputPort[14] (SplitPendulum_continuous_state) of System ::_ (MultibodyPlant<double>) is deprecated: Use the port suffxed with '_state' instead; the '_continuous_state' alias will be removed on or after 2023-03-01.
```

**Review by jwnimmer-tri — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/18274)* status: needs platform reviewer assigned, needs at least two assigned reviewers

___
*[`systems/framework/diagram.cc` line 518 at r1](https://reviewable.io/reviews/robotlocomotion/drake/18274#-NGDgUitC76v2yiUajGW:-NGDgUitC76v2yiUajGX:b-ydorlq) ([raw file](https://github.com/robotlocomotion/drake/blob/e9086823cd20f0b9be08ee15c1d682ab5e6f5777/systems/framework/diagram.cc#L518)):*
> ```cpp
>     const InputPortLocator& dest = edge.first;
>     const System<T>* dest_sys = dest.first;
>     src_sys->GetGraphvizOutputPortToken(src_sys->get_output_port(src.second),
> ```

FYI If you ask for a Graphviz display of a System or Diagram with deprecated ports, it will trigger the warning message.  I was satisfied with that effect.

**Comment by jwnimmer-tri** [comment]
+@RussTedrake for UX review, please -- do the API and semantics meet the requirements?

+@sherm1 for feature review, please -- are the API and semantics a good idea, and is the implementation sufficiently clear and well-tested?

+(status: do not merge) to be sure we remember to assign a platform reviewer later on.

**Review by RussTedrake — COMMENTED** [review]
Woohoo!  The OutputPort case looks fantastic.  Thanks!

Since you said in #12214 that you're not even sure that input port sugar is a net win, I gather that you think developers could start deprecating input ports immediately once this lands.  If yes, then can we add an example, e.g. in leaf_system_deprecation_test of how one might write an update or output method that properly uses the deprecated input port if it has values, or the new input port if it has values, and perhaps throws if they both have values?

For that matter, the example here of declaring the same output port twice, with one being deprecated, is great, but it will be removed on the deprecation timeline.  can we capture a version of that example in the test so it will persist?

> Reviewed 10 of 16 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/18274#-:-NGE66s4DIAVwvQmYSXg:b2ag69r)* status: LGTM missing from assignees sherm1(platform),RussTedrake(platform), labeled "do not merge" (waiting on @sherm1)

___
*[`multibody/plant/test/multibody_plant_deprecated_test.cc` line 34 at r1](https://reviewable.io/reviews/robotlocomotion/drake/18274#-NGE3DjTDGAi_nicd00T:-NGE3DjU29A1Aq6UF1ZV:b-oywjmu) ([raw file](https://github.com/robotlocomotion/drake/blob/e9086823cd20f0b9be08ee15c1d682ab5e6f5777/multibody/plant/test/multibody_plant_deprecated_test.cc#L34)):*
> ```cpp
>   EXPECT_EQ(all_state_port.size(), 6);
>   EXPECT_EQ(default_state_port.size(), 4);
>   EXPECT_EQ(pendulum_state_port.size(), 2);
> ```

nit: it would be slightly more convincing (and not much more work) to check the values are the same for the new and deprecated ports.  (probably for an arbitrary but non-default set of states)

**Review by jwnimmer-tri — COMMENTED** [review]
> I gather that you think developers could start deprecating input ports immediately once this lands.

Yes.  I think this is sufficient to deprecate both input and output ports.  I think the only missing thing would be any sugar for common-case ergonomics (e.g., renaming a port).

> If yes, then can we add an example, e.g. in leaf_system_deprecation_test of how one might write an update or output method that properly uses the deprecated input port if it has values, or the new input port if it has values, and perhaps throws if they both have values?

Good idea; done.

> For that matter, the example here of declaring the same output port twice, with one being deprecated, is great, but it will be removed on the deprecation timeline.  can we capture a version of that example in the test so it will persist?

Done.  The unit test demo has one input and one output.

> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/18274#-:-NGE_8hUE4Gz4UFBBcUj:b-eehlse)* status: LGTM missing from assignees sherm1(platform),RussTedrake(platform), labeled "do not merge" (waiting on @RussTedrake and @sherm1)

___
*[`multibody/plant/test/multibody_plant_deprecated_test.cc` line 34 at r1](https://reviewable.io/reviews/robotlocomotion/drake/18274#-NGE3DjTDGAi_nicd00T:-NGE_UK64F1lSPmyl7ui:b-qanu7h) ([raw file](https://github.com/robotlocomotion/drake/blob/e9086823cd20f0b9be08ee15c1d682ab5e6f5777/multibody/plant/test/multibody_plant_deprecated_test.cc#L34)):*
<details><summary><i>Previously, RussTedrake (Russ Tedrake) wrote…</i></summary><blockquote>

nit: it would be slightly more convincing (and not much more work) to check the values are the same for the new and deprecated ports.  (probably for an arbitrary but non-default set of states)
</blockquote></details>

Good catch.  I'd intended to do this, but I got distracted by the yak-shaving.

**Review by RussTedrake — COMMENTED** [review]
Awesome.  Thanks!

I didn't check all of the system implementation plumbing, but from the user perspective :lgtm_strong: 

> Reviewed 3 of 3 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/18274#-:-NGEb5ce4_AYE3v_sWew:b-2z5zdq)* status: LGTM missing from assignee sherm1(platform), labeled "do not merge" (waiting on @sherm1)

___
*[`systems/framework/test/leaf_system_deprecation_test.cc` line 246 at r2](https://reviewable.io/reviews/robotlocomotion/drake/18274#-NGEaptW7WQbELtMjjbs:-NGEaptW7WQbELtMjjbt:bhq26e2) ([raw file](https://github.com/robotlocomotion/drake/blob/97b4b2b80a0a73c04c4692e18d549767fe2a27ea/systems/framework/test/leaf_system_deprecation_test.cc#L246)):*
> ```cpp
>     const std::string message = "use the ports with suffix '_new'";
>     input_old_ = &this->DeclareVectorInputPort("input", 1);
>     output_old_ = &this->DeclareVectorOutputPort(
> ```

perhaps add a comment here helping people notice that we pass the same CalcOutput method to both the new port and the deprecated port?

**Review by sherm1 — COMMENTED** [review]
Feature pass done. A few comments:
- First, it is great to have a way to deprecate port names, however done -- thanks!
- The deprecation API looks OK, but I'm not wild about the downstream consequences (having to check multiple input ports for a single value)
- I think the message should go to the Console where it has a chance of being seen. Optionally you could log it as well but I don't think that matters much.
- The once-per-port-per-process logging is heroic, but I think too heavyweight for what it accomplishes. A user who is seeing deprecation messages should change their code asap (they will have to after the deprecation period anyway). In that case I don't think it is such a bad thing for them to get some redundant messages.
- Minor: the bool parameters are hard to follow at the call sites. I suggest either using two-valued enums with nice names or at least following every naked true/false with a `/*meaning*/` comment.

An alternative implementation we discussed a while back would be to allow port name deprecation aliases, which seems much simpler to me. I assume you rejected that approach for good reason but I can't remember why.

All that said, if you think this is the best approach I will lgtm it since it is clearly much better than not being able to deprecate these names at all!

> Reviewed 13 of 16 files at r1, 3 of 3 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/18274#-:-NGJNvRVBJD86MSBrZSg:bxmmzyq)* status: 5 unresolved discussions, LGTM missing from assignee sherm1(platform), labeled "do not merge" (waiting on @jwnimmer-tri)

___
*[`systems/framework/port_base.h` line 221 at r2](https://reviewable.io/reviews/robotlocomotion/drake/18274#-NGJ78-REvm2QsTSojp7:-NGJ78-REvm2QsTSojp8:byf4fpi) ([raw file](https://github.com/robotlocomotion/drake/blob/97b4b2b80a0a73c04c4692e18d549767fe2a27ea/systems/framework/port_base.h#L221)):*
> ```objc
>   }
> 
>   // Returns the bool stroed within the given port that indicates whether we've
> ```

typo stroed -> stored
___
*[`systems/framework/system.cc` line 46 at r2](https://reviewable.io/reviews/robotlocomotion/drake/18274#-NGJHvJ08u24NMcHi8bG:-NGJHvJ08u24NMcHi8bH:bu99knn) ([raw file](https://github.com/robotlocomotion/drake/blob/97b4b2b80a0a73c04c4692e18d549767fe2a27ea/systems/framework/system.cc#L46)):*
> ```cpp
>   DRAKE_THROW_UNLESS(input_port.get_data_type() == kVectorValued);
>   const InputPortBase& self_input_port_base = this->GetInputPortBaseOrThrow(
>       "AllocateInputVector", input_port.get_index(), false);
> ```

BTW `__func__` would be easier to grok than repeating the function name
___
*[`systems/framework/system_base.h` line 418 at r2](https://reviewable.io/reviews/robotlocomotion/drake/18274#-NGJNNudEFg6spidDh4t:-NGJNNudEFg6spidDh4u:b-9ph2b2) ([raw file](https://github.com/robotlocomotion/drake/blob/97b4b2b80a0a73c04c4692e18d549767fe2a27ea/systems/framework/system_base.h#L418)):*
> ```objc
>     DRAKE_DEMAND(0 <= index && index < num_input_ports());
>     if (input_ports_[index]->get_deprecation().has_value())
>       WarnPortDeprecation(true, index);
> ```

BTW I found it hard to keep track of the meaning of `true` and `false` at the various call sites. Would be nice here to see
`WarnPortDeprecation(kInputPort, index)` or something like that. Alternatively, 
`WarnPortDeprecation(true /*input port*/, index)` would be ok
___
*[`systems/framework/system_base.cc` line 357 at r2](https://reviewable.io/reviews/robotlocomotion/drake/18274#-NGJJm476YgdLEX8rb_J:-NGJJm476YgdLEX8rb_K:b4gtcob) ([raw file](https://github.com/robotlocomotion/drake/blob/97b4b2b80a0a73c04c4692e18d549767fe2a27ea/systems/framework/system_base.cc#L357)):*
> ```cpp
> 
>   // The had_already_warned above is a *per instance* warning, for performance.
>   // We'd like to warn at most onces *per process*; therefore, we have a second
> ```

typo onces -> once
___
*[`systems/framework/test/leaf_system_deprecation_test.cc` line 275 at r2](https://reviewable.io/reviews/robotlocomotion/drake/18274#-NGJMjFzEjr_R9zc3utQ:-NGJMjFzEjr_R9zc3utR:b21hy9d) ([raw file](https://github.com/robotlocomotion/drake/blob/97b4b2b80a0a73c04c4692e18d549767fe2a27ea/systems/framework/test/leaf_system_deprecation_test.cc#L275)):*
> ```cpp
>   auto context = dut.CreateDefaultContext();
> 
>   // The call to GetInputPort will log a warnning.
> ```

typo warnning

**Comment by jwnimmer-tri** [comment]
> The deprecation API looks OK, but I'm not wild about the downstream consequences (having to check multiple input ports for a single value)... An alternative implementation we discussed a while back would be to allow port name deprecation aliases, which seems much simpler to me. I assume you rejected that approach for good reason but I can't remember why.

#16269 and #16268 discussed it.  For one, it doesn't solve deprecation for _removing_ a port, only for _renaming_ it.  It also doesn't really do the right thing in case of exported subsystem ports within a Diagram which themselves need to inherit both port names (somehow).  It's simplest for the fundamental abstraction to be "every port has one name, period".  If we need the sugar for either-or input evaluation, we can add that atop without too much trouble.

> I think the message should go to the Console where it has a chance of being seen. Optionally you could log it as well but I don't think that matters much.

I thought that's what this PR is doing already.  Can you clarify?

> The once-per-port-per-process logging is heroic, but I think too heavyweight for what it accomplishes. A user who is seeing deprecation messages should change their code asap (they will have to after the deprecation period anyway). In that case I don't think it is such a bad thing for them to get some redundant messages.

If they are doing Monte-Carlo simulation, they could easy see 10,000+ copies of the warning.  I don't think that's OK.

**Review by jwnimmer-tri — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/18274)* status: 1 unresolved discussion, LGTM missing from assignee sherm1(platform), labeled "do not merge" (waiting on @RussTedrake and @sherm1)

___
*[`systems/framework/test/leaf_system_deprecation_test.cc` line 246 at r2](https://reviewable.io/reviews/robotlocomotion/drake/18274#-NGEaptW7WQbELtMjjbs:-NGJee1q0DaIVrZCzkyC:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/97b4b2b80a0a73c04c4692e18d549767fe2a27ea/systems/framework/test/leaf_system_deprecation_test.cc#L246)):*
<details><summary><i>Previously, RussTedrake (Russ Tedrake) wrote…</i></summary><blockquote>

perhaps add a comment here helping people notice that we pass the same CalcOutput method to both the new port and the deprecated port?
</blockquote></details>

Done.

**Review by jwnimmer-tri — COMMENTED** [review]
+@rpoyner-tri for platform review per schedule (tomorrow), please -(status: do not merge).

> Minor: the bool parameters are hard to follow at the call sites.

Yeah.  I didn't love it either.

The root cause for most of those is that we really want range-for iteration for all input ports (or output ports) but we don't have any function for that, and I mostly don't want to add any iterators stuff. (I'd rather just wait for C++ ranges before I bother with it.)

> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/18274#-:-NGJgKdX4pUXzi3UINoa:b-o3877g)* status: LGTM missing from assignees sherm1(platform),rpoyner-tri(platform), labeled "do not merge" (waiting on @rpoyner-tri, @RussTedrake, and @sherm1)

___
*[`systems/framework/system_base.h` line 418 at r2](https://reviewable.io/reviews/robotlocomotion/drake/18274#-NGJNNudEFg6spidDh4t:-NGJgBCM0q0tF48Rc0hX:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/97b4b2b80a0a73c04c4692e18d549767fe2a27ea/systems/framework/system_base.h#L418)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW I found it hard to keep track of the meaning of `true` and `false` at the various call sites. Would be nice here to see
`WarnPortDeprecation(kInputPort, index)` or something like that. Alternatively, 
`WarnPortDeprecation(true /*input port*/, index)` would be ok
</blockquote></details>

Done.

**Review by sherm1 — COMMENTED** [review]
>> I think the message should go to the Console where it has a chance of being seen. Optionally you could log it as well but I don't think that matters much.

>I thought that's what this PR is doing already. Can you clarify?

Oh, I assumed `drake::log()->warn()` went to a log file somewhere. If that goes to the console then nvm.

Feature :lgtm:

> Reviewed 4 of 6 files at r3, 5 of 5 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/18274#-:-NGJmA8n0LwsCMxjpQJb:bkyawee)* status: LGTM missing from assignee rpoyner-tri(platform) (waiting on @rpoyner-tri)

**Review by rpoyner-tri — COMMENTED** [review]
> Reviewed 2 of 2 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/18274)* status: :shipit: complete! all discussions resolved, LGTM from assignees rpoyner-tri(platform),sherm1(platform),RussTedrake(platform) (waiting on @jwnimmer-tri)
